lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,284 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'avro/io'
+require 'json'
+
+module Avro
+  module IO
+    # JsonDecoder keeps track of the current data and reads it using a given data type.
+    #
+    # About validations - If JSONDatumReader has its own read_data implementation, data validations can be moved there instead.
+    # That way reader can reuse Schema.validate method and have same behavior as the writer in case of schema and data mismatch.
+    class JsonDecoder
+      attr_reader :reader
+      attr_accessor :current_data
+
+      def initialize(reader)
+        @reader = reader
+
+        datum = @reader.read
+        datum = '""' if datum.empty?
+        data = JSON.parse(datum, {:quirks_mode => true})
+        @current_data = data
+      end
+
+      def read_null
+        error('null') unless current_data.nil?
+        nil
+      end
+
+      def read_boolean
+        error('boolean') unless Schema.is_boolean?(current_data)
+        current_data
+      end
+
+      def read_int
+        error('"int"') unless Schema.is_integer?(current_data)
```

Comment:
Otherwise for an invalid integer (`2147483648`),

JsonDatumWriter will throw this error during encoding - 
```
"The datum 2147483648 is not an example of schema \"int\""
```

JsonDatumReader will throw this error during decoding - 
```
"The datum 2147483648 is not an example of schema int"
```
Just wanted to be consistent in both cases.